### PR TITLE
Solved issue #353

### PR DESCRIPTION
**Pull request by Clemcarle**
Issue #353 has been solved!

### CODE REVIEW

**Comment by Clemcarle** [comment]
So this is quite mysterious, because it runs succesfully on my and Oliver's computers (Windows), but fails on Richel's and on github (Linux). Anyone has any idea what is going on?

**Comment by TheoPannetier** [comment]
This is odd indeed. I suspect this has to do with the RNG drawing different keys on different machines. Also the RNG itself might have changed since I wrote this test, so that the expected A is no longer drawn for either KAM. If so, I'll need to make a better test. I'll have a look sometime!

**Comment by richelbilderbeek** [comment]
I have checked the code. It is not odd that different key maps have the same key; it is only checked if *a* keymap has different keys :-)

**Comment by Clemcarle** [comment]
I am not sure about this, Richel: Making sure that different key maps can be generated will not have the same keys is the whole purpose of the `get_n_random_kams` function. Do you believe it is not doing that ?

**Comment by richelbilderbeek** [comment]
@Clemcarle you are right and I was wrong: I see on Linux that the key is in neither :-)

Screenshot from 2021-11-19 11-18-54

**Comment by TheoPannetier** [comment]
Ok, so my test is bad then: I assumed the same key would be drawn across all machines.
I'll come up with a better test 👍

**Comment by richelbilderbeek** [comment]
It is a fun failed test :-)

**Review by TheoPannetier — APPROVED** [review]
Brilliant work, I like your solution of drawing both random KAMs at once 👍

**Comment by TheoPannetier** [comment]
@richelbilderbeek what do you think of my new test for #353 ?
It's ugly because we run the test over 1000 semi-random iterations, but I don't see any other way to ensure that the test will work for any OS

**Comment by richelbilderbeek** [comment]
Hi @TheoPannetier, I like the test! Doing it once, however, is good enough here :-). 

A hint about that is the number of tests, 1000, of which the number probably already felt arbitrary :-)

File: game_options.cpp [hunk 1]
```diff
@@ -47,16 +47,19 @@ void music_on(game_options& o) noexcept
 
 game_options get_random_game_options(const int& rng_seed) {
   std::srand(rng_seed);
+  std::vector<key_action_map> random_kams = get_n_random_kams(2);
```
Comment (richelbilderbeek) [inline game_options.cpp:50]: I see why this is done. I suggest a comment such as:

```c++
// Need to create all KAMs at once to ensure all keys are unique over all KAMs
```

File: game_options.cpp [hunk 2]
```diff
@@ -147,18 +150,23 @@ void test_game_options()
...
-    const int rng_seed = 271; // by default this will assign A to both players
-    const game_options rgo = get_random_game_options(rng_seed);
-    key_action_map map_1 = rgo.get_kam_1();
-    key_action_map map_2 = rgo.get_kam_2();
-    bool a = map_1.has_key(sf::Keyboard::Key::A);
-    bool b = map_2.has_key(sf::Keyboard::Key::A);
-    // Key A is in KAM1 or KAM2 but not both
-    assert(a != b);
+    const int rng_seed = 271;
+    for (int seed = 0; seed < 1000; seed++)
```
Comment (richelbilderbeek) [inline game_options.cpp:158]: I enjoy the thoroughness of 1000 iterations! Still, this costs the project in the long run, as everytime the test suit is run, there are 999 superfluous iterations. Also, the feeling you got when picking 1000, i.e. 'Why 1000? Why no 100? Or 10? Or 10000?' is an indication that this was an arbitrary decision. 1 time is good enough here :-)

File: key_action_map.cpp [hunk 3]
```diff
@@ -106,9 +106,9 @@ key_action_map get_random_kam()
             new_key
             );
       if (key_match == v_random_keys.end())
-        {
+      {
```
Comment (richelbilderbeek) [inline key_action_map.cpp:109]: Yay, the boy scout rule!

File: key_action_map.cpp [hunk 4]
```diff
@@ -120,6 +120,43 @@ key_action_map get_random_kam()
       );
 }
 
+std::vector<key_action_map> get_n_random_kams(int n)
+{
+  std::vector<key_action_map> vector_kam;
```
Comment (richelbilderbeek) [inline key_action_map.cpp:125]: To be examplary, add `assert(n >= 0);`, else the for loop will do approx 2^64 too much iterations :-)

**Review by richelbilderbeek — CHANGES_REQUESTED** [review]
I enjoy seeing this work. I have some minor comments, but I feel they are important enough to pick 'Request Changes', also because I think you want to do excellent. Feel free to debate and ask me for the next review! Keep up the good job!

File: key_action_map.cpp (hunk 4 above)
Comment (TheoPannetier) [reply key_action_map.cpp:125]:
```suggestion
  assert(n >= 0);
  std::vector<key_action_map> vector_kam;
```

File: game_options.cpp (hunk 2 above)
Comment (TheoPannetier) [reply game_options.cpp]: Done!

**Review by richelbilderbeek — APPROVED** [review]
Happy to see you could agree with my proposed changes!

**Pull request merged**